lib/compress/zstd_compress_superblock.c

```diff
@@ -0,0 +1,567 @@
+ *  Stores literals block type (raw, rle, compressed) and
+ *  huffman description table to hufMetadata.
+ *  Currently, this does not consider the option of reusing huffman table from
+ *  previous super-block. I think it would be a good improvement to add that option.
+ *  @return : size of huffman description table or error code */
+static size_t ZSTD_buildSuperBlockEntropy_literal(void* const src, size_t srcSize,
+                                            const ZSTD_hufCTables_t* prevHuf,
+                                                  ZSTD_hufCTables_t* nextHuf,
+                                                  ZSTD_hufCTablesMetadata_t* hufMetadata)
+{
+    unsigned count[HUF_SYMBOLVALUE_MAX + 1];
+    unsigned maxSymbolValue = 255;
+    unsigned huffLog = 11;
+
+    DEBUGLOG(5, "ZSTD_buildSuperBlockEntropy_literal (srcSize=%zu)", srcSize);
+
+    /* Prepare nextEntropy assuming reusing the existing table */
+    memcpy(nextHuf, prevHuf, sizeof(*prevHuf));
+
+    /* small ? don't even attempt compression (speed opt) */
+#   define COMPRESS_LITERALS_SIZE_MIN 63
+    {   size_t const minLitSize = COMPRESS_LITERALS_SIZE_MIN;
+        if (srcSize <= minLitSize) { hufMetadata->hType = set_basic; return 0; }
+    }
+
+    /* Scan input and build symbol stats */
+    {   size_t const largest = HIST_count(count, &maxSymbolValue, (const BYTE*)src, srcSize);
+        FORWARD_IF_ERROR(largest);
+        if (largest == srcSize) { hufMetadata->hType = set_rle; return 0; }
+        if (largest <= (srcSize >> 7)+4) { hufMetadata->hType = set_basic; return 0; }
+    }
+
+
+    /* Build Huffman Tree */
+    memset(nextHuf->CTable, 0, sizeof(nextHuf->CTable));
+    huffLog = HUF_optimalTableLog(huffLog, srcSize, maxSymbolValue);
+    {   size_t const maxBits = HUF_buildCTable((HUF_CElt*)nextHuf->CTable, count,
+                                               maxSymbolValue, huffLog);
+        FORWARD_IF_ERROR(maxBits);
+        huffLog = (U32)maxBits;
+        hufMetadata->hType = set_compressed;
+        return HUF_writeCTable(hufMetadata->hufDesBuffer, sizeof(hufMetadata->hufDesBuffer),
+                              (HUF_CElt*)nextHuf->CTable, maxSymbolValue, huffLog);
+    }
+}
+
+/** ZSTD_buildSuperBlockEntropy_sequences() :
+ *  Builds entropy for the super-block sequences.
+ *  Stores symbol compression modes and fse table to fseMetadata.
+ *  @return : size of fse tables or error code */
+static size_t ZSTD_buildSuperBlockEntropy_sequences(seqStore_t* seqStorePtr,
+                                              const ZSTD_fseCTables_t* prevEntropy,
+                                                    ZSTD_fseCTables_t* nextEntropy,
+                                              const ZSTD_CCtx_params* cctxParams,
+                                                    ZSTD_fseCTablesMetadata_t* fseMetadata)
+{
+    ZSTD_strategy const strategy = cctxParams->cParams.strategy;
+    unsigned count[MaxSeq+1];
```

Comment:
This should also go in the workspace.

Reply:
I have replace everywhere with workspace by looking at zstd_compress to see which workspace is used for what purposes. This one, I am not sure which workspace to use since zstd_compress.c also uses stack memory for this purpose. This one is small though. Assuming unsigned is 4 byte this array is going to be 4 X 53 = 212 bytes.

Reply:
212 bytes seems reasonable to me.

If some very strict stack usage requirements mandate to use less, I believe it's better to make it a separate topic.

Reply:
I think count is allocated in stack because buildCTable() uses workspace and there is a danger of overwriting count. But I think buildCTable's workspace need is small enough that I think workspace can be used for both count[] and buildCTable. Let me try that. I am not sure how to make it future proof though. I will need some feedback.